agent/command/s3_copy.go

```diff
@@ -75,105 +95,140 @@ type s3Loc struct {
+func (s3pc *s3copy) ParseParams(params map[string]interface{}) error {
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		WeaklyTypedInput: true,
+		Result:           s3pc,
+	})
+	if err != nil {
+		return errors.WithStack(err)
 	}
-	if err := c.validateParams(); err != nil {
-		return errors.Wrapf(err, "error validating %v params", c.Name())
+
+	if err := decoder.Decode(params); err != nil {
+		return errors.Wrapf(err, "error decoding %s params", s3pc.Name())
 	}
-	return nil
+
+	return s3pc.validate()
 }
 
 // validateParams is a helper function that ensures all
 // the fields necessary for carrying out an S3 copy operation are present
-func (c *s3copy) validateParams() error {
-	if c.AwsKey == "" {
-		return errors.New("s3 AWS key cannot be blank")
+func (s3pc *s3copy) validate() error {
+	catcher := grip.NewSimpleCatcher()
+
+	// make sure the command params are valid
+	if s3pc.AwsKey == "" {
+		catcher.Add(errors.New("aws_key cannot be blank"))
 	}
-	if c.AwsSecret == "" {
-		return errors.New("s3 AWS secret cannot be blank")
+	if s3pc.AwsSecret == "" {
+		catcher.Add(errors.New("aws_secret cannot be blank"))
 	}
-	for _, s3CopyFile := range c.S3CopyFiles {
+
+	for _, s3CopyFile := range s3pc.S3CopyFiles {
+		// make sure both buckets are valid
+		if err := validateS3BucketName(s3CopyFile.Source.Bucket); err != nil {
+			return errors.Wrapf(err, "%v is an invalid bucket name", s3CopyFile.Source.Bucket)
+		}
+		if err := validateS3BucketName(s3CopyFile.Destination.Bucket); err != nil {
+			return errors.Wrapf(err, "%v is an invalid bucket name", s3CopyFile.Destination.Bucket)
+		}
+
 		if s3CopyFile.Source.Region == "" {
 			s3CopyFile.Source.Region = endpoints.UsEast1RegionID
 		}
 		if s3CopyFile.Destination.Region == "" {
 			s3CopyFile.Destination.Region = endpoints.UsEast1RegionID
 		}
-		if s3CopyFile.Source.Bucket == "" {
-			return errors.New("s3 source bucket cannot be blank")
-		}
-		if s3CopyFile.Destination.Bucket == "" {
-			return errors.New("s3 destination bucket cannot be blank")
-		}
-		if s3CopyFile.Source.Path == "" {
-			return errors.New("s3 source path cannot be blank")
-		}
-		if s3CopyFile.Destination.Path == "" {
-			return errors.New("s3 destination path cannot be blank")
-		}
-		if s3CopyFile.Permissions == "" {
-			s3CopyFile.Permissions = s3.BucketCannedACLPublicRead
-		}
+	}
 
-		err := validateS3BucketName(s3CopyFile.Source.Bucket)
-		if err != nil {
-			return errors.Wrapf(err, "source bucket '%v' is invalid",
-				s3CopyFile.Source.Bucket)
-		}
+	return catcher.Resolve()
+}
 
-		err = validateS3BucketName(s3CopyFile.Destination.Bucket)
-		if err != nil {
-			return errors.Wrapf(err, "destination bucket '%v' is invalid",
-				s3CopyFile.Destination.Bucket)
-		}
+// Apply the expansions from the relevant task config (including restricted expansion)
+// to all appropriate fields of the s3put.
+func (s3pc *s3copy) expandParams(conf *internal.TaskConfig) error {
```

Comment:
We're getting rid of restricted expansions, but anyway this command shouldn't be using them, only [s3.put and s3.get](https://github.com/evergreen-ci/evergreen/wiki/Project-and-Distro-Settings#variables). I'm not sure this helper function is really adding anything as opposed to just calling ExpandValues within Execute.